flang/lib/Optimizer/Dialect/FIROps.cpp

```diff
@@ -991,6 +991,23 @@ void fir::InsertOnRangeOp::build(mlir::OpBuilder &builder,
   build(builder, result, resTy, aggVal, eleVal, aa);
 }
 
+/// Range bounds must be nonnegative, and the range must not be empty.
+static mlir::LogicalResult verify(fir::InsertOnRangeOp op) {
+  bool rangeIsKnownToBeNonempty = false;
+  for (auto i = op.coor().end(), b = op.coor().begin(); i != b;) {
```

Comment:
Why traverse the list backwards? The ranges have to appear in pairs, whether the traversal is forward or backward, right?

It's not a bug, but it is a little surprising.

Reply:
It would be a bug to start from the front.  The traversal must start at the outermost dimension, which with column major order is the end.  In a dictionary order comparison, the first nonequal comparison is the one that determines the result.  Comparisons after that don't matter.  That is implemented here with the continue statement.